rules/msk_consume_in_module.go

```diff
@@ -0,0 +1,174 @@
+package rules
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/terraform-linters/tflint-plugin-sdk/hclext"
+	"github.com/terraform-linters/tflint-plugin-sdk/logger"
+	"github.com/terraform-linters/tflint-plugin-sdk/tflint"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// MskConsumeInModuleRule checks whether an MSK module only consumes from topics
+// defined in the module.
+type MskConsumeInModuleRule struct {
+	tflint.DefaultRule
+}
+
+func (r *MskConsumeInModuleRule) Name() string {
+	return "msk_consume_in_module"
+}
+
+func (r *MskConsumeInModuleRule) Enabled() bool {
+	return true
+}
+
+func (r *MskConsumeInModuleRule) Link() string {
+	return ReferenceLink(r.Name())
+}
+
+func (r *MskConsumeInModuleRule) Severity() tflint.Severity {
+	return tflint.ERROR
+}
+
+func (r *MskConsumeInModuleRule) Check(runner tflint.Runner) error {
+	path, err := runner.GetModulePath()
+	if err != nil {
+		return fmt.Errorf("getting module path: %w", err)
+	}
+	if !path.IsRoot() {
+		logger.Debug("skipping child module")
+		return nil
+	}
+
+	// resourceNameMap: resource_name -> topic_name (for mapping variables to EvalCtx)
+	// moduleTopics: topic_name -> struct{} (for name lookups)
+	resourceNameMap, moduleTopics, err := getKafkaTopics(runner)
+	if err != nil {
+		return err
+	}
+	logger.Debug("found topics", "topics", resourceNameMap)
+
+	modules, err := runner.GetModuleContent(
+		&hclext.BodySchema{
+			Blocks: []hclext.BlockSchema{
+				{
+					Type:       "module",
+					LabelNames: []string{"name"},
+					Body: &hclext.BodySchema{
+						Attributes: []hclext.AttributeSchema{
+							{Name: "produce_topics"},
+							{Name: "consume_topics"},
+						},
+					},
+				},
+			},
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("getting modules: %w", err)
+	}
+	evalCtx := buildTopicNameContext(resourceNameMap)
+	for _, block := range modules.Blocks {
+		for _, topicAttr := range []string{"consume_topics", "produce_topics"} {
+			if err := r.reportExternalTopics(runner, topicAttr, block, evalCtx, moduleTopics); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func getKafkaTopics(runner tflint.Runner) (map[string]string, map[string]struct{}, error) {
+	resourceContents, err := runner.GetResourceContent(
+		"kafka_topic",
+		&hclext.BodySchema{
+			Attributes: []hclext.AttributeSchema{{Name: "name", Required: true}},
+		},
+		nil,
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("getting kafka_topic contents: %w", err)
+	}
```

Comment:
:thinking: running on the terraform repo I see an error:

```
TFLint in dev-aws/kafka-shared-msk/pubsub/:
Failed to check ruleset; failed to check "msk_consume_in_module" rule: getting kafka_topic contents: pubsub-test-throughput.tf:24,54-54: Missing required argument; The argument "name" is required, but no definition was found., and 17 other diagnostic(s)
```

But that block is a `kafka_acl` https://github.com/utilitywarehouse/kafka-cluster-config/blob/d28b50424f9d38d31ddceb50338df62a6934694f/dev-aws/kafka-shared-msk/pubsub/pubsub-test-throughput.tf#L24 I thought by passing `"kafka_topic"` as the first argument to `GetResourceContent` the runner would _only_ try and parse those blocks (otherwise what's the point of that argument?)

Reply:
Indeed, I would expect that too. Looking at the implementation, it should do that: https://github.com/terraform-linters/tflint-plugin-sdk/blob/master/plugin/internal/plugin2host/client.go#L80

Reply:
Just skipping it :shrug: 711ec0edda0b55107762d7da215fffb1ac5f91e6